### PR TITLE
Add custom dimension row serializer

### PR DESCRIPTION
**Pull request by garyluoex**
so that user don't have to override toString method in order for it to work

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Serialization strategy for DimensionRow.
+ */
+public class DimensionRowSerializer extends JsonSerializer<DimensionRow> {
+
+    @Override
+    public void serialize(DimensionRow value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
```
Comment (QubitPi) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java]: Why do we throw exception at method level?

**Review by ghost — CHANGES_REQUESTED** [review]

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy [hunk 2]
```diff
@@ -12,18 +14,25 @@ class DimensionRowSpec extends Specification {
...
         testRow.getRowMap() == [(nonKeyFieldName): nonKeyValue, (keyFieldFieldName): keyValue]
     }
+
+    def "json serialization into the correct format we expect"() {
+        setup:
+        ObjectMapper objectMapper = new ObjectMapper()
+
+        expect:
+        objectMapper.writeValueAsString(testRow) == '{"nonKeyFieldName":"nonKey","keyFieldName":"key"}'
+    }
```
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy]: Is this test resilient about map key order?  Are we guaranteeing that anywhere?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,6 +10,9 @@ Current
 
 ### Added:
 
+- [Add custom dimension row serializer](https://github.com/yahoo/fili/pull/399)
+    * Add custom dimension row serializer so that dimension row serialize into dimension row cache correctly
+
```
Comment (michael-mclawhorn) [inline CHANGELOG.md]: spelling: 'dimension rows serialize' or 'each dimension row serializes'

**Review by michael-mclawhorn — COMMENTED** [review]
Nothing much at issue here.

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java (hunk 1 above)
Comment (garyluoex) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java]:
```java    
public abstract void serialize(T value, JsonGenerator gen, SerializerProvider serializers)
        throws IOException, JsonProcessingException;
```

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2016 Yahoo Inc.
```
Comment (kevinhinterlong) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java]: 2017

File: fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java (hunk 1 above)
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/DimensionRowSerializer.java]: My total bad.  I'm not sure if I was looking at the wrong javadoc, but I see in 2.2 for the fasterxml code, these are thrown.
I swear that I saw a version somewhere where there were not explicit throws.

**Review by michael-mclawhorn — APPROVED** [review]
👍

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,9 @@ Current
 
 ### Added:
 
+- [Add custom dimension row serializer](https://github.com/yahoo/fili/pull/399)
+    * Add custom dimension row serializer so that each dimension row serializes into dimension row cache correctly
+
```
Comment (michael-mclawhorn) [inline CHANGELOG.md]: I suggest you move this to a fix and describe that this resolves where dimension field names do not serialize correctly.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy [hunk 6]
```diff
@@ -12,18 +14,27 @@ class DimensionRowSpec extends Specification {
...
-        setup:
-        DimensionRow testRow = new DimensionRow(keyField, [(keyField): keyValue, (nonKeyField): nonKeyValue])
-
         expect:
         testRow.getKeyValue() == keyValue
         testRow.getRowMap() == [(nonKeyFieldName): nonKeyValue, (keyFieldFieldName): keyValue]
     }
+
+    def "json serialization into the correct format we expect"() {
+        setup:
```
Comment (michael-mclawhorn) [inline fili-core/src/test/groovy/com/yahoo/bard/webservice/data/dimension/DimensionRowSpec.groovy]: 'correct format we expect' is redundant.  
Perhaps 'all expected fields and values are present in the dimension row serialization'

**Review by michael-mclawhorn — COMMENTED** [review]
My approval stands, but I've added a few more suggestions for improvement.

**Comment by michael-mclawhorn** [comment]
@QubitPi could you rereview please and update any comments?

**Pull request closed without merging**